clang/test/CheckedC/inferred-bounds/bounds-context.c

```diff
@@ -1857,29 +1857,80 @@ void inc_dec_bounds4(array_ptr<int> a : bounds(a, a)) { // expected-note {{(expa
   // CHECK-NEXT: }
 }
 
-// Increment/decrement operators on non-variables or variables without declared bounds
-// do not result in bounds checking-related warnings or errors
+// Increment/decrement operators on dereference expressions or array subscripts
+// of type _Nt_array_ptr<T> are bounds checked
+// Increment/decrement operators on variables without declared bounds are not
+// bounds checked
 void inc_dec_bounds5(nt_array_ptr<int> *p, array_ptr<int> a) {
-  // Observed bounds context after increment:  { }
-  ++*p;
+  // Observed bounds context after increment:  { *p => bounds(*p - 1, (*p - 1) + 0) }
+  ++*p; // expected-warning {{cannot prove declared bounds for '*p' are valid after increment}} \
```

Comment:
I think this behavior of the compiler is incorrect. Perhaps, at present, we should restrict the bounds validation for `*p` and `p[i]` when the type of `*p` or `p[i]` is a checked pointer type?

Reply:
In this particular test case, `*p` has type `_Nt_array_ptr<int>`, since `p` has type `_Nt_array_ptr<int> *`.

In general, if `e` is a pointer dereference `*e1` or an array subscript `e1[e2]`, we will only check the bounds of `e` if `e` fulfills one of two conditions:

1. `e` has type `_Ptr<T>`. In this case, the target bounds of `e` are `bounds((_Array_ptr<T>)e, (_Array_ptr<T>)e + 1)`.
2. `e` has type `_Nt_array_ptr<T>`. In this case, the target bounds of `e` are `bounds(e, e + 0)`.

We do not explicitly check the type of `e` in `UpdateAfterAssignment` or `ValidateBoundsContext`. Instead, the `AbstractSet` containing `e` is added to `ObservedBounds` if and only if `e` has target bounds, which will only happen in the two cases described above.

The methods `UnaryOperatorTargetBounds` and `ArraySubscriptExprTargetBounds` in SemaBounds.cpp include comments that state that "Currently, we don't know the target bounds of a pointer stored in a pointer dereference (or returned by a subscripting operation), unless it is a _Ptr type or an _Nt_array_ptr."

Reply:
Thanks for the clarification (I missed noticing the `*`)!